README.md

```diff
@@ -145,17 +165,16 @@ code, include a script or makefile that will do the job.
 
 ## A2.5
 
-«replace this with your answer»
-
+files to test are parser.h, parser.cpp, main.cpp
+run main.cpp and it will input true and false based on tests declared in the main
```

Comment:
First off, thank you for writing both positive and negative tests. My only comment is that they're not really tests: they just run code out output stuff, leaving it for a poor human to work out whether the code is good. A unit test will run the code and report if the value returned is not the expected value.

I quickly hacked your code:

~~~ c++
#include <iostream>
#include <assert.h>

#include "parser.h"
typedef vector<string> sentence;
typedef vector<sentence> sentences;

using namespace std;
int main(){
    sentences valid {
      {"the", "lazy", "cat", "ate", "slowly", "EOI"},
      {"the", "lazy", "dog", "ate", "slowly", "EOI"},
      {"the", "smelly", "cat", "ran", "slowly", "EOI"},
      {"the", "smelly", "dog", "ran", "noisily", "EOI"},
      {"the", "smelly", "cat", "ate", "noisily", "EOI"},
    };

    sentences invalid {
      {"lazy", "cat", "ate", "slowly", "EOI"},
      {"dog", "ate", "slowly", "EOI"},
      {"the", "smelly", "ran", "slowly", "EOI"},
      {"the", "smelly", "dog", "ran", "noisily"},
      {"EOI"},
    };

    parser parse;

    for (sentence s: valid) {
        assert(parse.FSManalyze(s));
    }

    for (sentence s: invalid) {
        assert(!parse.FSManalyze(s));
    }
}
~~~

# -2

On to the parser.

First, I have to say that your code was a joy to read. I've had a lifetime's worth of C++ in the last two days, and this was this chunk of code that didn't make a part of my soul curl up a little. So 3 bonus points from my soul.

# +3

However, you actually didn't do what the assignment wanted: I was hoping you'd take the table from above and transcode it into a C++ data structure (`map<map<int,string>, int>`) and then work from that. At the same time, you did actually use states in your code, so rather than deducting 6 for this, I'm deducting 3

# -3

FYI: here's the code using a table

~~~ cpp
#include <iostream>
using namespace std;
#include <map>

typedef int state;
typedef std::map<std::string, state> transitions;
typedef std::map<state, transitions> state_table;

const int END = 99;

state_table states = {
  { 0, {{ "the", 1 }}},
	{ 1, {{ "lazy", 2}, {"smelly", 2}, {"dog", 3}, {"cat", 3}}},
	{ 2, {{ "lazy", 2}, {"smelly", 2}, {"dog", 3}, {"cat", 3}}},
	{ 3, {{ "ate", 4},  {"ran",    4}}},
	{ 4, {{ "slowly", 5}, { "noisily", 5}, {"EOI", END}}},
	{ 5, {{ "EOI", END }}}
};


int getNextState(int currentState, string nextWord) {
	transitions t = states[currentState];
	if (t.count(nextWord))
		return t[nextWord];
	else
		return -1;
}
~~~